pkg/dataplane/deployment/os.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployment
+
+import (
+	"context"
+
+	"k8s.io/apiserver/pkg/storage/names"
+
+	dataplaneutil "github.com/openstack-k8s-operators/dataplane-operator/pkg/util"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	ansibleeev1alpha1 "github.com/openstack-k8s-operators/openstack-ansibleee-operator/api/v1alpha1"
+)
+
+// ConfigureOS ensures the node Operating System config
+func ConfigureOS(ctx context.Context, helper *helper.Helper, namespace string, sshKeySecret string, inventoryConfigMap string) error {
+
+	role := ansibleeev1alpha1.Role{
+		Name:           "Deploy EDPM Operating System Configure",
+		Hosts:          "all",
+		Strategy:       "free",
+		Become:         true,
+		AnyErrorsFatal: true,
+		Tasks: []ansibleeev1alpha1.Task{
+			{
+				Name: "Configure edpm_podman",
```

Comment:
I would refactor this function to avoid duplicating the block:

```
{
    Name: "Configure edpm_podman",
    ImportRole: ansibleeev1alpha1.ImportRole{
        Name:      "<ROLE_NAME>",
        TasksFrom: "<ROLE_YAML_FILE_WITH_TASKS>",
     },
    Tags: []string{"<TAGS>"},
},
```

and provide a data structure (e.g. a `map`) as input of this function that you can iterate, producing the blocks you need.
Not 100% convinced the right way is to do the `OS` part this way, I would have created a playbook containing all the roles added here, and called it from this point w/ a set of `tags/variables` to properly drive its execution, but if you prefer to keep this approach is fine, just some optimizations are required I guess

Reply:
I don't think a map would provide too much value,   we would be moving the duplication from `ansibleeev1alpha1.Task` to `map[string]string` 
```go
var importRole = []map[string]string {
        {
	  "ROLE_NAME": "name",
          "ROLE_YAML_FILE_WITH_TASKS": "yaml",
          "TAGS": "tag1, tag2",
        },
       {
	  "ROLE_NAME": "name2",
          "ROLE_YAML_FILE_WITH_TASKS": "yaml2",
          "TAGS": "tag1, tag2",
        },
    }
  ```
  
  About creating a playbook, do you mean something like this? https://github.com/openstack-k8s-operators/openstack-ansibleee-operator/blob/main/examples/openstack-ansibleee-play.yaml#L8-L14

Reply:
> I don't think a map would provide too much value, we would be moving the duplication from `ansibleeev1alpha1.Task` to `map[string]string`
> 
> ```go
> var importRole = []map[string]string {
>         {
> 	  "ROLE_NAME": "name",
>           "ROLE_YAML_FILE_WITH_TASKS": "yaml",
>           "TAGS": "tag1, tag2",
>         },
>        {
> 	  "ROLE_NAME": "name2",
>           "ROLE_YAML_FILE_WITH_TASKS": "yaml2",
>           "TAGS": "tag1, tag2",
>         },
>     }
> ```
> 
> About creating a playbook, do you mean something like this? https://github.com/openstack-k8s-operators/openstack-ansibleee-operator/blob/main/examples/openstack-ansibleee-play.yaml#L8-L14

As you wish, in general I prefer working with a more complex data structure instead of making the code verbose and with a lot of duplication, but I guess it's just an opinionated approach.
e.g.

```
type Task struct {
	Name string
	Path string
	Tags string
}

	tx := Task { Name:"podman", Path: "podman.yaml", Tags: []string{"podman", "edpm_os"} }
	...
	...

	var v = map[string]Tu {
		...
                ...
		"taskX": tx,
		...
		...
	}

```
What I meant regarding the playbook is that calling something like [1], where `roles` are imported on a file like [1] (which is just an example) allows you to just call an `os.yaml` as a playbook instead of building it in a Go function.


[1] https://github.com/openstack/tripleo-ansible/blob/master/tripleo_ansible/playbooks/cli-deployed-ceph.yaml#L191-L287

Reply:
Gotcha! I'm following the pattern @slagle started with network https://github.com/openstack-k8s-operators/dataplane-operator/pull/33 
I believe we should have the same pattern for all the playbooks: https://github.com/openstack-k8s-operators/edpm-ansible/tree/main/edpm_ansible/playbooks
@slagle what do you think about the playbook approach?

Reply:
I'm seeing other playbooks that have tasks that use ansible modules: https://github.com/openstack-k8s-operators/edpm-ansible/blob/main/edpm_ansible/playbooks/deploy-edpm-os-install.yml
So going with playbook seems to be the best way

FYI: @slagle @fultonj @rebtoor

Reply:
I'm fine with any approach to avoid the code duplication. It's really just a list of role names and tasks_from.

As for playbooks vs roles, I'd tend to avoid playbooks. We want to orchestrate from the operator, and be able to configure how we call the roles. If we used playbooks, we'd have to use playbook templates anyway. This will allow us to do more advanced things later on, like executing roles in parallel (multiple AnsibleEE's). I'm not saying it will *never* be playbooks, or other more specific configurations like ceph can't use playbooks, just that as of now, I'd like to drive it all with the existing standalone roles for the tasks that dataplane-operator is directly orchestrating.

For the tasks in the playbooks from edpm-ansible, we just need to move those into an existing standalone role in edpm-ansible. For the ones from os-install in particular, I'd just copy them straight into the edpm_bootstrap role.

Reply:
I'm moving them: https://github.com/openstack-k8s-operators/edpm-ansible/pull/45